jekyll/_posts/2021-03-12-new-Nim-channels.md

```diff
@@ -0,0 +1,233 @@
+---
+
+title: "The new Nim channels implementation for ORC"
+author: xflywind
+excerpt: "The new channels library is efficient and safe to use"
+
+---
+
+# The new Nim channels implementation for ORC
+
+Version 1.4 ships with the so-called ORC memory management algorithm. ORC is the existing ARC algorithm (first shipped in version 1.2) plus a cycle collector. The Nim devel branch also introduces a new module called `std/isolation`. With it we can pass `isolated` data to threads safely and easily. It prevents data races at compile time. Recently `std/channels` is merged to the devel branch which is designed for ORC. It combines `isolated` data and `channels` and is efficient and safe to use.
+
+**Note:** you need the Nim devel branch to compile the code below.
+
+## Background
+
+A channel is a model for sharing memory via message passing. A thread is able to send or receive messages over a channel. It's like sending a letter to your friend. The postman is the channel. Your friend is the receiver.  You may know  `system/channels` already exists. What’s the difference between new channels implementation and the old one? If you use the old one, you need to copy your letter by hand first and send the copied one to your friend instead. Then your friend may mark something on the copied letter, it won’t affect the original letter. It works fine, however it is not efficient. If you use the new one, you only need to put your letter in the mailbox. No need to copy your letter!
+
+## The advantages
+
+- Designed for ARC/ORC, no legacy code
+- No need to `deepcopy`, just move data around
+- No data races
+- Based on [Michael & Scott lock-based queues](https://www.cs.rochester.edu/~scott/papers/1996_PODC_queues.pdf)
+
+## Explore the new channels
+
+**Note:** Be sure to compile your code with `--gc:orc –-threads:on`.
+
+### Let's crawl the web
+
+**todo_urls.json**
+
+```json
+{"url": ["https://google.com", "https://nim-lang.org"]}
+```
+
+**app.nim**
+
+The main thread prepares tasks via reading `todo_urls.txt`. Then it sends JSON data to a channel. The crawl thread does the crawlers’ work. It receives URL data from the channel and downloads the contents using `httpclient`.
+
+```nim
+import std/channels
+import std/[httpclient, isolation, json]
+
+
+var ch = initChan[JsonNode]() # we need to send JsonNode
+
+proc download(client: HttpClient, url: string) =
+  let response = client.get(url)
+  echo "content: "
+  echo response.body[0 .. 20]  # prints the results
+
+proc crawl =
+  var client = newHttpClient() # the crawler
+  var data: JsonNode
+  ch.recv(data) # the JSON data
+  if data != nil:
```

Comment:
can data ever be nil?
```
nim r --eval:'import std/json; echo parseJson("{}") == nil'
false
```

Reply:
If `recv` doesn't modify data, it will be nil.